File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const T = Number(input[0])
const answer = []

for (let tc = 0; tc < T; tc++) {
  const W = input[tc * 2 + 1]
  const K = Number(input[tc * 2 + 2])
  let MAX = Number.MIN_SAFE_INTEGER
  let MIN = Number.MAX_SAFE_INTEGER

  // k개 포함된 문자열
  // 1. K개 포함된 알파벳만 구하기
  // 2. 그 중에서 가장 긴 문자열 길이 구하기
  // 시작과 끝이 같은 알파벳이여야함

  const alphabet = new Map()

  for (let i = 0; i < W.length; i++) {
    const char = W[i]
    if (!alphabet.has(char)) {
      alphabet.set(char, [])
    }

    alphabet.get(char).push(i)
  }

  const candidate = Array.from(alphabet).filter(([_, values]) => values.length >= K)

  for (let [char, values] of candidate) {
    for (let i = 0; i <= values.length - K; i += 1) {
      MAX = Math.max(MAX, values[i + K - 1] - values[i] + 1)
      MIN = Math.min(MIN, values[i + K - 1] - values[i] + 1)
    }
  }
  if (MIN == Number.MAX_SAFE_INTEGER) {
    answer.push("-1")
  } else {
    answer.push(`${MIN} ${MAX}`)
  }
}
console.log(answer.join("\n"))

File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.py
import sys
from collections import defaultdict
input = sys.stdin.readline

# 입력 처리
T = int(input().strip())
answer = []

for tc in range(T):
    w = list(input().strip())
    k = int(input().strip())
    
    # 정확히 k개 포함하는 문자들 종류
    kind_of_char = defaultdict(list)
    
    for index, value in enumerate(w):
        kind_of_char[value].append(index)
    
    # 필터링: k개 이상 포함하는 문자들
    filtered_keys = {key: value for key, value in kind_of_char.items() if len(value) >= k}
    
    MAX = float('-inf')
    MIN = float('inf')
    
    for key, value in filtered_keys.items():
        for idx in range(len(value) - k + 1):
            MAX = max(MAX, value[idx + k - 1] - value[idx] + 1)
            MIN = min(MIN, value[idx + k - 1] - value[idx] + 1)
    
    if MAX == float('-inf') or MIN == float('inf'):
        answer.append("-1")
    else:
        answer.append(f"{MIN} {MAX}")

# 결과 출력
print("\n".join(answer))
